File: sdks/GoudEngine/Components/Sprite.cs
using System;
using System.Runtime.InteropServices;
using GoudEngine.Math;

namespace GoudEngine.Components
{
    /// <summary>
    /// 2D sprite component for rendering textured quads.
    /// </summary>
    /// <remarks>
    /// This is a pure data struct. All sprite manipulation logic is implemented in
    /// the Rust engine and accessible via FFI functions (goud_sprite_*).
    /// 
    /// Sprites are textured rectangles that can be tinted, flipped, and have custom
    /// UV coordinates for sprite sheets.
    /// 
    /// Note: Memory layout must match Rust FfiSprite.
    /// </remarks>
    [StructLayout(LayoutKind.Sequential)]
    [Component(0xFEDCBA0987654321, 48)] // TODO: Compute actual TypeId hash
    public struct Sprite : IComponent
    {
        /// <summary>
        /// Handle to the texture asset. Value of 0 means no texture (will render as colored quad).
        /// </summary>
        public ulong TextureHandle;

        /// <summary>
        /// Color tint red component (0.0 - 1.0).
        /// </summary>
        public float ColorR;

        /// <summary>
        /// Color tint green component (0.0 - 1.0).
        /// </summary>
        public float ColorG;

        /// <summary>
        /// Color tint blue component (0.0 - 1.0).
        /// </summary>
        public float ColorB;

        /// <summary>
        /// Color tint alpha component (0.0 - 1.0).
        /// </summary>
        public float ColorA;

        /// <summary>
        /// Source rectangle X position (if HasSourceRect is true).
        /// </summary>
        public float SourceRectX;

        /// <summary>
        /// Source rectangle Y position.
        /// </summary>
        public float SourceRectY;

        /// <summary>
        /// Source rectangle width.
        /// </summary>
        public float SourceRectWidth;

        /// <summary>
        /// Source rectangle height.
        /// </summary>
        public float SourceRectHeight;

        /// <summary>
        /// Whether source_rect is set.
        /// </summary>
        [MarshalAs(UnmanagedType.U1)]
        public bool HasSourceRect;

        /// <summary>
        /// If true, flip the sprite horizontally.
        /// </summary>
        [MarshalAs(UnmanagedType.U1)]
        public bool FlipX;

        /// <summary>
        /// If true, flip the sprite vertically.
        /// </summary>
        [MarshalAs(UnmanagedType.U1)]
        public bool FlipY;

        /// <summary>
        /// Anchor point X in normalized coordinates (0.0-1.0).
        /// </summary>
        public float AnchorX;

        /// <summary>
        /// Anchor point Y in normalized coordinates (0.0-1.0).
        /// </summary>
        public float AnchorY;

        /// <summary>
        /// Custom size width (if HasCustomSize is true).
        /// </summary>
        public float CustomSizeX;

        /// <summary>
        /// Custom size height.
        /// </summary>
        public float CustomSizeY;

        /// <summary>
        /// Whether custom_size is set.
        /// </summary>
        [MarshalAs(UnmanagedType.U1)]
        public bool HasCustomSize;

        /// <inheritdoc/>
        public ulong TypeId => 0xFEDCBA0987654321; // TODO: Compute actual hash

        /// <inheritdoc/>
        public int SizeInBytes => 48;

        /// <summary>
        /// Creates a new Sprite with the specified texture handle and default settings.
        /// </summary>
        /// <param name="textureHandle">Handle to the texture asset (0 for none).</param>
        public Sprite(ulong textureHandle)
        {
            TextureHandle = textureHandle;
            ColorR = 1.0f;
            ColorG = 1.0f;
            ColorB = 1.0f;
            ColorA = 1.0f;
            SourceRectX = 0.0f;
            SourceRectY = 0.0f;
            SourceRectWidth = 0.0f;
            SourceRectHeight = 0.0f;
            HasSourceRect = false;
            FlipX = false;
            FlipY = false;
            AnchorX = 0.5f;
            AnchorY = 0.5f;
            CustomSizeX = 0.0f;
            CustomSizeY = 0.0f;
            HasCustomSize = false;
        }

        /// <summary>
        /// Creates a new white sprite (no texture).
        /// </summary>
        public static Sprite Default() => new(0);

        /// <summary>
        /// Gets or sets the color tint.
        /// </summary>
        public Color Color
        {
            readonly get => new(ColorR, ColorG, ColorB, ColorA);
            set
            {
                ColorR = value.R;
                ColorG = value.G;
                ColorB = value.B;
                ColorA = value.A;
            }
        }

        /// <summary>
        /// Gets or sets the anchor point as a Vector2 (normalized 0.0-1.0).
        /// </summary>
        public Vector2 Anchor
        {
            readonly get => new(AnchorX, AnchorY);
            set
            {
                AnchorX = value.X;
                AnchorY = value.Y;
            }
        }

        /// <summary>
        /// Checks if the sprite is flipped on either axis.
        /// </summary>
        public readonly bool IsFlipped => FlipX || FlipY;

        /// <inheritdoc/>
        public override readonly string ToString()
        {
            var parts = new System.Collections.Generic.List<string>
            {
                $"tex: {TextureHandle}",
                $"color: RGBA({ColorR:F2}, {ColorG:F2}, {ColorB:F2}, {ColorA:F2})"
            };

            if (HasSourceRect)
            {
                parts.Add($"src: ({SourceRectX}, {SourceRectY}, {SourceRectWidth}x{SourceRectHeight})");
            }

            if (FlipX) parts.Add("flipX");
            if (FlipY) parts.Add("flipY");

            if (HasCustomSize)
            {
                parts.Add($"size: ({CustomSizeX}, {CustomSizeY})");
            }

            return $"Sprite({string.Join(", ", parts)})";
        }
    }

    /// <summary>
    /// RGBA color with components in the range [0.0, 1.0].
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct Color
    {
        public float R;
        public float G;
        public float B;
        public float A;

        public Color(float r, float g, float b, float a = 1.0f)
        {
            R = r;
            G = g;
            B = b;
            A = a;
        }

        public static readonly Color White = new(1, 1, 1, 1);
        public static readonly Color Black = new(0, 0, 0, 1);
        public static readonly Color Red = new(1, 0, 0, 1);
        public static readonly Color Green = new(0, 1, 0, 1);
        public static readonly Color Blue = new(0, 0, 1, 1);
        public static readonly Color Yellow = new(1, 1, 0, 1);
        public static readonly Color Transparent = new(0, 0, 0, 0);

        public override readonly string ToString() => $"RGBA({R:F2}, {G:F2}, {B:F2}, {A:F2})";
    }

    /// <summary>
    /// Rectangle with position and size.
    /// </summary>
    [StructLayout(LayoutKind.Sequential)]
    public struct Rect
    {
        public float X;
        public float Y;
        public float Width;
        public float Height;

        public Rect(float x, float y, float width, float height)
        {
            X = x;
            Y = y;
            Width = width;
            Height = height;
        }

        public override readonly string ToString() => $"Rect({X}, {Y}, {Width}x{Height})";
    }
}

File: goud_engine/src/core/types/sprite.rs
//! FFI-safe sprite and collision contact types.

// =============================================================================
// Contact Type
// =============================================================================

/// FFI-compatible contact information from a collision.
#[repr(C)]
#[derive(Debug, Clone, Copy, Default)]
pub struct GoudContact {
    /// Contact point X coordinate
    pub point_x: f32,
    /// Contact point Y coordinate
    pub point_y: f32,
    /// Collision normal X component (unit vector pointing from A to B)
    pub normal_x: f32,
    /// Collision normal Y component
    pub normal_y: f32,
    /// Penetration depth (positive = overlapping)
    pub penetration: f32,
}

impl From<crate::ecs::collision::Contact> for GoudContact {
    fn from(contact: crate::ecs::collision::Contact) -> Self {
        Self {
            point_x: contact.point.x,
            point_y: contact.point.y,
            normal_x: contact.normal.x,
            normal_y: contact.normal.y,
            penetration: contact.penetration,
        }
    }
}

// =============================================================================
// Sprite Types
// =============================================================================

/// FFI-safe Sprite representation.
///
/// This is a simplified version of the Sprite component suitable for FFI.
/// It uses raw u64 texture handles instead of generic AssetHandle types.
#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct FfiSprite {
    /// Texture handle (index and generation packed as u64).
    pub texture_handle: u64,

    /// Color tint red component (0.0 - 1.0).
    pub color_r: f32,
    /// Color tint green component (0.0 - 1.0).
    pub color_g: f32,
    /// Color tint blue component (0.0 - 1.0).
    pub color_b: f32,
    /// Color tint alpha component (0.0 - 1.0).
    pub color_a: f32,

    /// Source rectangle X position (if has_source_rect is true).
    pub source_rect_x: f32,
    /// Source rectangle Y position.
    pub source_rect_y: f32,
    /// Source rectangle width.
    pub source_rect_width: f32,
    /// Source rectangle height.
    pub source_rect_height: f32,
    /// Whether source_rect is set.
    pub has_source_rect: bool,

    /// Flip horizontally flag.
    pub flip_x: bool,
    /// Flip vertically flag.
    pub flip_y: bool,

    /// Anchor point X (normalized 0-1).
    pub anchor_x: f32,
    /// Anchor point Y (normalized 0-1).
    pub anchor_y: f32,

    /// Custom size width (if has_custom_size is true).
    pub custom_size_x: f32,
    /// Custom size height.
    pub custom_size_y: f32,
    /// Whether custom_size is set.
    pub has_custom_size: bool,
}

/// Heap-allocated sprite builder for FFI use.
///
/// This builder allows constructing a sprite by setting properties one at
/// a time without copying the entire struct on each modification.
#[repr(C)]
pub struct FfiSpriteBuilder {
    /// The sprite being built.
    pub sprite: FfiSprite,
}
